### PR TITLE
Add more options for TPU profiling

### PR DESCRIPTION
**Pull request by qiuminxu**
* Motivation for features / changes
StartTracing API has more options to config how to profile a model. Add that to the capture profile pop up.
* Technical description of changes
1) Add work_list, num_tracing_attempts and include_dataset_ops to the pop up.
2) Trim the string of the ProfilerServiceAddress to remove leading and trailing spaces.
3) Add auto-validate to the input field.
4) Fix a bug that the "_profilerServiceAddress" should be passed into the shouldDisableCaptureProfileButton instead of the "profilerServiceAddress".

* Screenshots of UI changes
capture_profile_button

* Detailed steps to verify changes work correctly (as executed by you)
bazel run tensorboard/plugins/profile:profile_demo
bazel run tensorboard:tensorboard -- --logdir=/tmp/profile_demo
* Alternate designs / implementations considered

### CODE REVIEW

File: tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html [hunk 1]
```diff
@@ -81,11 +83,34 @@ <h3>Profiling isn’t supported in Colab yet.</h3>
...
+    <paper-input label="Automatically retry N times when no trace event is collected"
+                 always-float-label
+                 type="number"
+                 min=1
+                 max=100
+                 auto-validate
+                 pattern="[0-9]+"
+                 error-message="Input is not an integer between 1 to 100"
+                 required
+                 value="{{_profileNumTracingAttempts}}"></paper-input>
```
Comment (qqfish) [inline tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: This option should be useful for GPU as well

File: tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html [hunk 2]
```diff
@@ -81,11 +83,34 @@ <h3>Profiling isn’t supported in Colab yet.</h3>
...
                type="number"
                min=1
+               max=600000
+               auto-validate
+               pattern="[0-9]+"
+               error-message="Input is not an integer between 1 to 600000"
+               required
                value="{{_profileDuration}}"></paper-input>
+  <div hidden="{{!_shouldShowTpuOptions(_profilerServiceAddressType)}}">
+    <paper-checkbox checked="{{_profileIncludeDatasetOps}}">Trace dataset ops</paper-checkbox>
```
Comment (qqfish) [inline tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: What are the differences between dataset ops and other cpu activities? Why do we need to turn this on explicitly for TPU? I think for GPU, all dataset ops will be traced once turn on cpu profiling.

Comment (qiuminxu) [reply tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: The dataset ops was found to be dominant in the CPU traces, removing that allows us to see more TPU device events. As you said, the filtering was not done for GPU, therefore I set it as an TPU only option.

File: tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html (hunk 1 above)
Comment (qiuminxu) [reply tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: Done. Moved that to show in GPU capture profile page as well.

File: tensorboard/plugins/profile/profile_plugin.py [hunk 3]
```diff
@@ -427,18 +427,16 @@ def data_route(self, request):
   @wrappers.Request.application
   def capture_route(self, request):
     service_addr = request.args.get('service_addr')
-    try:
-      duration = int(request.args.get('duration'))
-    except TypeError:
-      # Returns code=200 to show error message in UI.
-      return http_util.Respond(request, {'error': 'invalid duration.'},
-                               'application/json', code=200)
+    duration = int(request.args.get('duration'))
```
Comment (wchargin) [inline tensorboard/plugins/profile/profile_plugin.py]: Is `duration` intended to be an optional query parameter? If so, you’ll
need to provide a default here (`request.args.get('duration', '0')`), or
the `int(None)` conversion will be a type error. If it’s intended to be
mandatory, please use `request.args['duration']` to make that clear.

File: tensorboard/plugins/profile/profile_plugin.py [hunk 4]
```diff
@@ -427,18 +427,16 @@ def data_route(self, request):
...
     is_tpu_name = request.args.get('is_tpu_name') == 'true'
-    workers_list = ''
+    worker_list = request.args.get('worker_list')
+    include_dataset_ops = request.args.get('include_dataset_ops') == 'true'
+    num_tracing_attempts = int(request.args.get('num_retry')) + 1
```
Comment (wchargin) [inline tensorboard/plugins/profile/profile_plugin.py]: As with `duration` above.

File: tensorboard/plugins/profile/profile_plugin.py [hunk 5]
```diff
@@ -447,16 +445,17 @@ def capture_route(self, request):
...
       # on which profiler service is running.
       master_ip = master_grpc_addr.replace('grpc://', '').replace(':8470', '')
       service_addr = master_ip + ':8466'
       # Set the master TPU for streaming trace viewer.
       self.master_tpu_unsecure_channel = master_ip
     try:
-      profiler_client.start_tracing(service_addr, self.logdir,
-                                    duration, workers_list);
+      profiler_client.start_tracing(service_addr, self.logdir, duration,
+          worker_list, include_dataset_ops, num_tracing_attempts);
```
Comment (wchargin) [inline tensorboard/plugins/profile/profile_plugin.py]: nit: Superfluous semicolon on this line.

File: tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html [hunk 6]
```diff
@@ -388,6 +415,17 @@ <h3>No profile data was found.</h3>
         type: Number,
         value: 1000,
       },
+      _profileNumRetry: {
+        type: Number,
```
Comment (wchargin) [inline tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: Here `_profileNumRetry` is declared as a number, but below you compare
it for strict equality with `""`. It seems like one of these must be
wrong—does the `paper-input` provide a string or a number?

File: tensorboard/plugins/profile/profile_plugin.py (hunk 3 above)
Comment (qiuminxu) [reply tensorboard/plugins/profile/profile_plugin.py]: Done, set a default duration to 1s.

File: tensorboard/plugins/profile/profile_plugin.py (hunk 4 above)
Comment (qiuminxu) [reply tensorboard/plugins/profile/profile_plugin.py]: Done.

File: tensorboard/plugins/profile/profile_plugin.py (hunk 5 above)
Comment (qiuminxu) [reply tensorboard/plugins/profile/profile_plugin.py]: Done, thanks for spotting this.

File: tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html (hunk 6 above)
Comment (qiuminxu) [reply tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: Yes, it is String. Also fixed for _profileDuration

File: tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html [hunk 7]
```diff
@@ -385,9 +412,20 @@ <h3>No profile data was found.</h3>
         value: "ip-address",
       },
       _profileDuration: {
-        type: Number,
+        type: String,
         value: 1000,
       },
+      _profileNumRetry: {
+        type: String,
+        value: 3,
```
Comment (wchargin) [inline tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: If `type` is `String`, then `value` should be `"3"`, not `3`, right?

(Likewise for `_profileDuration` above.)

Comment (qiuminxu) [reply tensorboard/plugins/profile/tf_profile_dashboard/tf-profile-dashboard.html]: Yep, done, thanks :)

**Pull request merged**